idaes/core/scaling/autoscaling.py

```diff
@@ -0,0 +1,291 @@
+            ValueError if norm is not a positive integer
+        """
+        # Validate norm
+        # First cast to int
+        norm = int(norm)
+        if norm < 1:
+            raise ValueError(
+                f"Invalid value for norm in constraints_by_jacobian_norm ({norm}). "
+                "Value must be a positive integer."
+            )
+
+        if isinstance(blk_or_cons, BlockData):
+            # Scalar Block or element of Indexed Block
+            if descend_into:
+                # Scale all constraints, so pass con_list=None
+                con_list = None
+            else:
+                # Otherwise, get an iterator of constraints only on block
+                con_list = blk_or_cons.component_data_objects(
+                    Constraint, descend_into=False
+                )
+            self._con_by_norm(blk_or_cons, con_list=con_list, norm=norm)
+
+        elif isinstance(blk_or_cons, Block):
+            # Indexed Block
+            for b in blk_or_cons.values():
+                if descend_into:
+                    # Scale all constraints, so pass con_list=None
+                    con_list = None
+                else:
+                    # Otherwise, get an iterator of constraints only on block
+                    con_list = b.component_data_objects(Constraint, descend_into=False)
+                self._con_by_norm(b, con_list=con_list, norm=norm)
+
+        elif isinstance(blk_or_cons, ConstraintData):
+            # Scalar Constraint or element of Indexed Constraint
+            self._con_by_norm(
+                blk_or_cons.parent_block(), con_list=[blk_or_cons], norm=norm
+            )
+
+        elif isinstance(blk_or_cons, Constraint):
+            # Indexed Constraint
+            self._con_by_norm(
+                blk_or_cons.parent_block(), con_list=blk_or_cons.values(), norm=norm
+            )
+
+        else:
+            raise TypeError(f"{blk_or_cons.name} is not a block or constraint.")
+
+    def _vardata_by_magnitude(self, vardata):
+        if vardata.fixed:
+            # Fixed var
+            return
+
+        if value(vardata) is None:
+            sf = 1
+        else:
+            val = abs(value(vardata))
+            if val <= self.config.zero_tolerance:
+                sf = 1
+            else:
+                sf = 1 / val
+
+        self._set_scaling_factor(vardata, "variable", sf)
+
+    def _con_by_norm(self, blk, con_list=None, norm=2):
+        # Get scaled Jacobian - we want to consider any existing scaling
+        # We will account for existing scaling factors later and update them
+        jac, nlp = get_jacobian(blk, scaled=True)
+
+        if con_list is None:
+            con_list = nlp.get_pyomo_equality_constraints()
+
+        # Use scipy to get all the norms
+        # Should be more efficient that iterating in Python
+        axis = (
+            1  # Could make this an argument to also support variable-based norm scaling
```

Comment:
Based on your comment, should we consider adding another method for variable-based norm scaling?

Reply:
I do not know if this makes sense or not - comments would be welcome (@dallan-keylogic ).

Reply:
If we scale variables first based on order-of-magnitude, we can then scale constraints based on row/column norm. However, norm-based variable scaling probably won't be useful for a fully unscaled model, because the most common Jacobian entries will be O(1). 

For a partially-scaled model, though, filling in variable scaling factors based on column norms might be useful.